server/lib/src/idm/server.rs

```diff
@@ -1275,127 +1276,133 @@ impl<'a> IdmServerAuthTransaction<'a> {
+                let mut softlock_write = self.softlocks.write();
+                let slock = Arc::new(Mutex::new(CredSoftLock::new(cred_slock_policy)));
+                softlock_write.insert(cred_id, slock.clone());
+                softlock_write.commit();
+                slock
             }
-        } else {
-            None
         };
 
-        // Validate the unix_pw - this checks the account/cred lock states.
-        let res = if let Some(mut slock) = maybe_valid {
-            // Account is unlocked, can proceed.
-            account
-                .verify_unix_credential(uae.cleartext.as_str(), &self.async_tx, ct)
-                .inspect(|res| {
-                    if res.is_none() {
-                        // Update it.
-                        slock.record_failure(ct);
-                    };
-                })
-        } else {
-            // Account is slocked!
+        let mut slock = slock_ref.lock().await;
+
+        slock.apply_time_step(ct);
+
+        if !slock.is_valid() {
             security_info!("Account is softlocked.");
-            Ok(None)
+            //TODO: Did we want to return an error condition?
+            return Ok(None);
+        }
+
+        let Ok(password) = cred.password_ref() else {
+            // The credential should only ever be a password
+            unreachable!();
         };
-        res
-    }
 
-    pub async fn token_auth_ldap(
-        &mut self,
-        lae: &LdapTokenAuthEvent,
-        ct: Duration,
-    ) -> Result<Option<LdapBoundToken>, OperationError> {
-        match self.validate_and_parse_token_to_token(&lae.token, ct)? {
-            Token::UserAuthToken(uat) => {
-                let spn = uat.spn.clone();
-                Ok(Some(LdapBoundToken {
-                    session_id: uat.session_id,
-                    spn,
-                    effective_session: LdapSession::UserAuthToken(uat),
-                }))
-            }
-            Token::ApiToken(apit, entry) => {
-                let spn = entry
-                    .get_ava_single_proto_string(Attribute::Spn)
-                    .ok_or_else(|| {
-                        OperationError::InvalidAccountState("Missing attribute: spn".to_string())
-                    })?;
+        // Check the provided password against the stored hash
+        let valid = password.verify(cleartext).map_err(|e| {
+            error!(crypto_err = ?e);
+            e.into()
+        })?;
 
-                Ok(Some(LdapBoundToken {
-                    session_id: apit.token_id,
-                    spn,
-                    effective_session: LdapSession::ApiToken(apit),
+        if !valid {
+            // Update it.
+            slock.record_failure(ct);
+
+            return Ok(None);
+        }
+
+        security_info!("Successfully authenticated with unix (or primary) password");
+        if password.requires_upgrade() {
+            //TODO: Do we really want to throw if pushing to the async queue fails?
```

Comment:
If the queue fails here it indicates a pretty catastrophic failure of the backend. It also will impact the ability for sessions to be created or other security data to be updated. So  Ithink that the error here is right, we should "fail noisy".

Reply:
Right, So the queue is integral to the function of kanidm, I thought it was a async task list of "stuff we gotta do at some point" and therefore non critical

Reply:
Yeah, it's for allowing concurrent things to write back "eventually", and potentially batched. A really good example here is auth session data. If a user logs in, we async write the session data to the DB in the bg. When the session token is issued there is a short grace window (5 mins) to account for replication delay etc, where the session token will still validate if the session isn't in the db yet. 

But after that if the session isn't in the db, the session is canned. 

So here if the session is sent for async write back and it fails, then any session this server issues will fail to work. That's a pretty big failure for us.

Similar, the async pw upgrade writeback is to "raise" the security level/quality of a password. If that fails, that means we could be holding on to insecure pw material when we shouldn't.

Anyway, better to error here. I think it's better to have "1 narrow path to success" where everything has to work, and to make it fragile to get there. And then we error at the slightest chance of something being wrong, because we don't want unintended consequences.

Reply:
Probably a smart choice given that this does authentication. I am still in the mindset of "make the user experience as pleasant as possible". 

pw upgrade queue could be handled gracefully since it would just requeue the next time that user logs in. But if the same queue is also responsible for writing sessions into the database, then its failure to write is a big enough deal to warrant stopping an execution flow.